easybuild/framework/easyconfig/tweak.py

```diff
@@ -843,16 +963,126 @@ def map_easyconfig_to_target_tc_hierarchy(ec_spec, toolchain_mapping, targetdir=
             if dep_tc_name in toolchain_mapping:
                 orig_dep['toolchain'] = toolchain_mapping[dep_tc_name]
             # Replace the binutils version (if necessary)
+            dep_changed = False
             if 'binutils' in toolchain_mapping and (dep['name'] == 'binutils' and dep_tc_name == GCCcore.NAME):
                 orig_dep.update(toolchain_mapping['binutils'])
-                # set module names
+                dep_changed = True
+            elif update_dep_versions:
+                # Search for available updates for this dependency:
+                # First get highest version candidate paths for this (include search through subtoolchains)
+                potential_version_matches = find_potential_version_mappings(dep, toolchain_mapping,
+                                                                            versonsuffix_mapping=versonsuffix_mapping)
+                # Only highest version match is retained by default in potential_version_matches, compare that version
+                # to the original version and replace if appropriate (upgrades only).
+                if potential_version_matches:
+                    highest_version_match = potential_version_matches[0]['version']
+                    if LooseVersion(highest_version_match) > LooseVersion(dep['version']):
+                        _log.info("Updating version of %s dependency from %s to %s", dep['name'], dep['version'],
+                                  highest_version_match)
+                        _log.info("Depending on your configuration, this will be resolved with one of the following "
+                                  "easyconfigs: \n%s", '\n'.join(cand['path'] for cand in potential_version_matches))
+                        orig_dep['version'] = highest_version_match
+                        if orig_dep['versionsuffix'] in versonsuffix_mapping:
+                            dep['versionsuffix'] = versonsuffix_mapping[orig_dep['versionsuffix']]
+                            orig_dep['versionsuffix'] = versonsuffix_mapping[orig_dep['versionsuffix']]
```

Comment:
I am struggling to find any reason why we do :)

Reply:
Ah, I took this approach from somewhere else in the code base. I don't understand what this does
```
        for idx, dep in enumerate(parsed_ec['ec'][key]):

            # reference to original dep dict, this is the one we should be updating
            orig_dep = parsed_ec['ec']._config[key][0][idx]
```
so I assumed it is somehow necessary

Reply:
I think it is necessary because one of them is parsed and the other isn't but you know better than me

Reply:
I tried removing `orig_dep` and you do need both, you need the parsed deps for the MNS

Reply:
What do you mean with "parsed deps" exactly? Where template values are resolved?
Where is the MNS relevant here?

Reply:
Maybe I should edit that comment :) My understanding is that `dep` has been parsed but `orig_dep` has not and `orig_dep` is the one that is relevant to the `dump` method which is why we need that target specifically. I assumed it must be related to the MNS because below the MNS is called to determine the module name using `dep` and not `orig_dep`.

The truth is, I don't know what is wrong but it doesn't work correctly if you replace `dep` by `orig_dep` (and modify the name to just `dep`), the only places where this can have an impact is
```
dep_tc_name = dep['toolchain']['name']  # unlikely
potential_version_matches = find_potential_version_mappings(dep, toolchain_mapping, versonsuffix_mapping=versonsuffix_mapping)
LooseVersion(dep['version'])  # unlikely
orig_dep['short_mod_name'] = ActiveMNS().det_short_module_name(dep)
orig_dep['full_mod_name'] = ActiveMNS().det_full_module_name(dep)
```

There is a material difference in the output with the following case if you try to make the switch (at least when I tried it):
```
eb -D --robot --try-toolchain=foss,2018b --minimal-toolchains --disable-cleanup-tmpdir --try-update-deps Keras-2.1.3-intel-2017b-Python-3.6.3.eb
```

Reply:
Just to confirm, I redid this change and there are many dependencies that are not updated in the case where `orig_dep` and `dep` are one and the same. Output difference from command
```
eb -D --robot --try-toolchain=foss,2018b --minimal-toolchains --disable-cleanup-tmpdir --try-update-deps Keras-2.1.3-intel-2017b-Python-3.6.3.eb
```
is
```
20c21
<  * [ ] /home/alanc/EasyBuild_Git/easybuild-easyconfigs/easybuild/easyconfigs/l/libyaml/libyaml-0.1.7-GCCcore-7.3.0.eb (module: Compiler/GCCcore/7.3.0 | libyaml/0.1.7)
>  * [ ] /home/alanc/EasyBuild_Git/easybuild-easyconfigs/easybuild/easyconfigs/l/libyaml/libyaml-0.2.1-GCCcore-7.3.0.eb (module: Compiler/GCCcore/7.3.0 | libyaml/0.2.1)
40a40,41
>  * [ ] /home/alanc/EasyBuild_Git/easybuild-easyconfigs/easybuild/easyconfigs/p/protobuf/protobuf-3.6.1-GCCcore-7.3.0.eb (module: Compiler/GCCcore/7.3.0 | protobuf/3.6.1)
50,56c51,57
<  * [ ] /tmp/eb-2afs_g/tweaked_dep_easyconfigs/Theano-1.0.1-foss-2018b-Python-3.6.6.eb (module: MPI/GCC/7.3.0-2.30/OpenMPI/3.1.1 | Theano/1.0.1-Python-3.6.6)
<  * [ ] /tmp/eb-2afs_g/tweaked_dep_easyconfigs/wheel-0.30.0-foss-2018b-Python-3.6.6.eb (module: MPI/GCC/7.3.0-2.30/OpenMPI/3.1.1 | wheel/0.30.0-Python-3.6.6)
<  * [ ] /tmp/eb-2afs_g/tweaked_dep_easyconfigs/PyYAML-3.12-foss-2018b-Python-3.6.6.eb (module: MPI/GCC/7.3.0-2.30/OpenMPI/3.1.1 | PyYAML/3.12-Python-3.6.6)
<  * [ ] /tmp/eb-2afs_g/tweaked_dep_easyconfigs/pkgconfig-1.2.2-foss-2018b-Python-3.6.6.eb (module: MPI/GCC/7.3.0-2.30/OpenMPI/3.1.1 | pkgconfig/1.2.2-Python-3.6.6)
<  * [ ] /tmp/eb-2afs_g/tweaked_dep_easyconfigs/TensorFlow-1.5.0-foss-2018b-Python-3.6.6.eb (module: MPI/GCC/7.3.0-2.30/OpenMPI/3.1.1 | TensorFlow/1.5.0-Python-3.6.6)
<  * [ ] /tmp/eb-2afs_g/tweaked_dep_easyconfigs/h5py-2.7.1-foss-2018b-Python-3.6.6.eb (module: MPI/GCC/7.3.0-2.30/OpenMPI/3.1.1 | h5py/2.7.1-Python-3.6.6)
<  * [ ] /tmp/eb-2afs_g/tweaked_easyconfigs/Keras-2.1.3-foss-2018b-Python-3.6.6.eb (module: MPI/GCC/7.3.0-2.30/OpenMPI/3.1.1 | Keras/2.1.3-Python-3.6.6)
---
>  * [ ] /home/alanc/EasyBuild_Git/easybuild-easyconfigs/easybuild/easyconfigs/t/Theano/Theano-1.0.3-foss-2018b-Python-3.6.6.eb (module: MPI/GCC/7.3.0-2.30/OpenMPI/3.1.1 | Theano/1.0.3-Python-3.6.6)
>  * [ ] /home/alanc/EasyBuild_Git/easybuild-easyconfigs/easybuild/easyconfigs/w/wheel/wheel-0.31.1-foss-2018b-Python-3.6.6.eb (module: MPI/GCC/7.3.0-2.30/OpenMPI/3.1.1 | wheel/0.31.1-Python-3.6.6)
>  * [ ] /home/alanc/EasyBuild_Git/easybuild-easyconfigs/easybuild/easyconfigs/p/PyYAML/PyYAML-3.13-foss-2018b-Python-3.6.6.eb (module: MPI/GCC/7.3.0-2.30/OpenMPI/3.1.1 | PyYAML/3.13-Python-3.6.6)
>  * [ ] /home/alanc/EasyBuild_Git/easybuild-easyconfigs/easybuild/easyconfigs/p/pkgconfig/pkgconfig-1.3.1-foss-2018b-Python-3.6.6.eb (module: MPI/GCC/7.3.0-2.30/OpenMPI/3.1.1 | pkgconfig/1.3.1-Python-3.6.6)
>  * [ ] /home/alanc/EasyBuild_Git/easybuild-easyconfigs/easybuild/easyconfigs/h/h5py/h5py-2.8.0-foss-2018b-Python-3.6.6.eb (module: MPI/GCC/7.3.0-2.30/OpenMPI/3.1.1 | h5py/2.8.0-Python-3.6.6)
>  * [ ] /home/alanc/EasyBuild_Git/easybuild-easyconfigs/easybuild/easyconfigs/t/TensorFlow/TensorFlow-1.12.0-foss-2018b-Python-3.6.6.eb (module: MPI/GCC/7.3.0-2.30/OpenMPI/3.1.1 | TensorFlow/1.12.0-Python-3.6.6)
>  * [ ] /tmp/eb-neWxA8/tweaked_easyconfigs/Keras-2.1.3-foss-2018b-Python-3.6.6.eb (module: MPI/GCC/7.3.0-2.30/OpenMPI/3.1.1 | Keras/2.1.3-Python-3.6.6)
```

Reply:
It's not so much parsed vs not parsed, it has to do with the templating.

`dep` is a *copy* of the original dependency, with templates resolved; `orig_dep` is a ref to the original dependency

We may need to update both in this context to make things work correctly. Perhaps updating `dep` isn't strictly needed, but it's cheap, and we've spent enough time on this... ;)